hazelcast/include/hazelcast/client/spi/impl/AbstractInvocationFuture.h

```diff
@@ -0,0 +1,550 @@
+                          public boost::enable_shared_from_this<AbstractInvocationFuture<T> > {
+                public:
+                    virtual ~AbstractInvocationFuture() {
+                    }
+
+                    virtual void andThen(const boost::shared_ptr<ExecutionCallback<T> > &callback) {
+                        andThen(callback, defaultExecutor);
+                    }
+
+                    virtual void andThen(const boost::shared_ptr<ExecutionCallback<T> > &callback,
+                                         const boost::shared_ptr<Executor> &executor) {
+                        Preconditions::isNotNull(callback, "callback");
+                        Preconditions::isNotNull(executor, "executor");
+
+                        boost::shared_ptr<BaseState> waiter(new ExecutionCallbackState(callback));
+                        boost::shared_ptr<BaseState> response = registerWaiter(waiter, executor);
+                        if (response != VOIDOBJECT) {
+                            unblock(callback, executor);
+                        }
+                    }
+
+                    virtual bool cancel(bool mayInterruptIfRunning) {
+                        return complete(boost::shared_ptr<IException>(
+                                new concurrent::CancellationException("AbstractInvocationFuture::cancel",
+                                                                      "Task was cancelled.")));
+                    }
+
+                    virtual bool isCancelled() {
+                        const boost::shared_ptr<BaseState> currentState = state.get();
+                        return (currentState->getType() == BaseState::Exception &&
+                                boost::static_pointer_cast<ExceptionState>(
+                                        currentState)->getException()->getErrorCode() == protocol::CANCELLATION);
+                    }
+
+                    virtual bool isDone() {
+                        return isDone(state);
+                    }
+
+                    virtual boost::shared_ptr<T> get() {
+                        boost::shared_ptr<ThreadState> thread = boost::shared_ptr<ThreadState>(
+                                new ThreadState(util::getCurrentThreadId()));
+                        boost::shared_ptr<BaseState> response = registerWaiter(thread, boost::shared_ptr<Executor>());
+                        if (response != VOIDOBJECT) {
+                            // no registration was done since a value is available.
+                            return resolveAndThrowIfException(response);
+                        }
+
+                        for (;;) {
+                            thread->park();
+                            if (isDone()) {
+                                return resolveAndThrowIfException(state);
+                            }
+                        }
+                    }
+
+                    virtual boost::shared_ptr<T> get(int64_t timeout, const TimeUnit &unit) {
```

Comment:
As far as I see this (time-timeunit) pair is used a lot. It is not the scope of this PR but would it be better to have a class for this so that it would be easier to pass it around and do operations on it.

Reply:
why isnt timeout also a constant?

Reply:
I thought making it constant can speed up things as the compiler can make some optimizations if it is constant. So even though it is copied it will be faster.

See `https://stackoverflow.com/questions/16033751/c-const-modifier-with-primitive-types`